File: FresherAndroid/AndroidBasicCode/ManhPT14_IVI_BAD_Practice_T01/app/src/main/java/com/example/contactapp/database/dao/ContactDao.java
package com.example.contactapp.database.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.example.contactapp.model.Contact;

import java.util.List;

@Dao
public interface ContactDao {

    @Insert
    long insert(Contact contact);

    @Update
    int update(Contact contact);

    @Delete
    int delete(Contact contact);

    // Lấy tất cả, sắp xếp theo alphabet
    @Query("SELECT * FROM contacts ORDER BY name COLLATE NOCASE ASC")
    List<Contact> getAll();

    @Query("SELECT * FROM contacts WHERE id = :id LIMIT 1")
    Contact getById(int id);

    // Tìm theo tên hoặc số điện thoại
    @Query("SELECT * FROM contacts WHERE name LIKE :kw OR phone LIKE :kw ORDER BY name COLLATE NOCASE ASC")
    List<Contact> search(String kw);

    @Query("SELECT COUNT(*) FROM contacts")
    int count();
}

File: FresherAndroid/AndroidAdvanced/submit/ManhPT14_ReactiveProgramming/app/src/main/java/com/example/reactivecontactapp/data/local/ContactDao.kt
package com.example.reactivecontactapp.data.local

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable


@Dao
interface ContactDao {

    @Query("SELECT * FROM contacts ORDER BY name ASC")
    fun getAllContacts(): Observable<List<Contact>>

    @Query("SELECT * FROM contacts WHERE name LIKE :query OR phone LIKE :query ORDER BY name ASC")
    fun search(query: String): Observable<List<Contact>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(contact: Contact): Completable

    @Update
    fun update(contact: Contact): Completable

    @Delete
    fun delete(contact: Contact): Completable
}